crates/bevy_ecs/src/entity/allocator.rs

```diff
@@ -0,0 +1,966 @@
+    }
+
+    /// Iterates the entities in these indices.
+    ///
+    /// # Safety
+    ///
+    /// [`Self::set`] must have been called on these indices before to initialize memory.
+    #[inline]
+    unsafe fn iter(&self, indices: core::ops::Range<u32>) -> FreeBufferIterator {
+        FreeBufferIterator {
+            buffer: self,
+            future_buffer_indices: indices,
+            current_chunk_slice: [].iter(),
+        }
+    }
+}
+
+impl Drop for FreeBuffer {
+    fn drop(&mut self) {
+        for index in 0..Self::NUM_CHUNKS {
+            let capacity = Self::capacity_of_chunk(index);
+            // SAFETY: we have `&mut` and the capacity is correct.
+            unsafe { self.0[index as usize].dealloc(capacity) };
+        }
+    }
+}
+
+/// An iterator over a [`FreeBuffer`].
+///
+/// # Safety
+///
+/// [`FreeBuffer::set`] must have been called on these indices beforehand to initialize memory.
+struct FreeBufferIterator<'a> {
+    buffer: &'a FreeBuffer,
+    /// The part of the buffer we are iterating at the moment.
+    current_chunk_slice: core::slice::Iter<'a, Slot>,
+    /// The indices in the buffer that are not yet in `current_chunk_slice`.
+    future_buffer_indices: core::ops::Range<u32>,
+}
+
+impl<'a> Iterator for FreeBufferIterator<'a> {
+    type Item = Entity;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(found) = self.current_chunk_slice.next() {
+            return Some(found.get_entity());
+        }
+
+        let still_need = self.future_buffer_indices.len() as u32;
+        let next_index = self.future_buffer_indices.next()?;
+        let (chunk, index, chunk_capacity) = self.buffer.index_in_chunk(next_index);
+
+        // SAFETY: Assured by constructor
+        let slice = unsafe { chunk.get_slice(index, still_need, chunk_capacity) };
+        self.future_buffer_indices.start += slice.len() as u32;
+        self.current_chunk_slice = slice.iter();
+
+        // SAFETY: Constructor ensures these indices are valid in the buffer; the buffer is not sparse, and we just got the next slice.
+        // So the only way for the slice to be empty is if the constructor did not uphold safety.
+        let next = unsafe { self.current_chunk_slice.next().debug_checked_unwrap() };
+        Some(next.get_entity())
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.future_buffer_indices.len() + self.current_chunk_slice.len();
+        (len, Some(len))
+    }
+}
+
+impl<'a> ExactSizeIterator for FreeBufferIterator<'a> {}
+impl<'a> core::iter::FusedIterator for FreeBufferIterator<'a> {}
+
+/// This tracks the state of a [`FreeCount`], which has lots of information packed into it.
+///
+/// - The first 33 bits store a signed 33 bit integer. This behaves like a u33, but we define `1 << 32` as 0.
+/// - The 34th bit stores a flag that indicates if the count has been disabled/suspended.
```

Comment:
To avoid ambiguity, it would probably be better to use "least significant bits" and "most significant bits" rather than "first bits" and "last bits". Just a nit.